pytorch_lightning/loops/dataloader/evaluation_loop.py

```diff
@@ -271,17 +271,86 @@ def _on_evaluation_epoch_end(self) -> None:
         self.trainer._call_lightning_module_hook("on_epoch_end")
         self.trainer.logger_connector.on_epoch_end()
 
+    def _get_keys(self, iterable: Union[List, dict]) -> Iterable[str]:
+        if isinstance(iterable, list):
+            for i in iterable:
+                yield from self._get_keys(i)
+        elif isinstance(iterable, dict):
+            for k, v in iterable.items():
+                if isinstance(v, dict):
+                    yield from self._get_keys(v)
+                else:
+                    yield k
+
+    def _find_value(self, data: dict, target: str) -> Iterable[Any]:
+        for k, v in data.items():
+            if k == target:
+                yield v
+            elif isinstance(v, dict):
+                yield from self._find_value(v, target)
+
     def _print_results(self, results: List[_OUT_DICT], stage: RunningStage) -> None:
-        # TODO: this could be updated to look nicer
-        from pprint import pprint
-
-        print("-" * 80)
-        for i, metrics_dict in enumerate(results):
-            print(f"DATALOADER:{i} {stage.upper()} RESULTS")
-            pprint(
-                {
-                    k: (v.item() if v.numel() == 1 else v.tolist()) if isinstance(v, torch.Tensor) else v
-                    for k, v in metrics_dict.items()
-                }
-            )
-            print("-" * 80)
+        if _RICH_AVAILABLE:
+            from rich.console import Console
+            from rich.table import Table
+
+            console = Console()
+            table = Table()
+
+            table.add_column("Metric", justify="center", style="cyan")
+
+            for i, result in enumerate(results):
+                clean_keys = [i.split("/dataloader_idx_")[0] for i in result.keys()]
+                results[i] = dict(zip(clean_keys, result.values()))
+
+            unique_keys: List[str] = sorted(list(set(self._get_keys(results))))
+            rows: List[List[str]] = [[i] for i in unique_keys]
+
+            for i, metrics in enumerate(results):
+                table.add_column(f"DATALOADER {i}", justify="center", style="magenta")
+
+                for metric in rows:
+                    v = list(self._find_value(metrics, metric[0]))
+                    if v:
+                        metric.append(f"{v[0]}")
+                    else:
+                        metric.append(" ")
+
+            for row in rows:
+                table.add_row(*row)
+
+            console.print("", table)
+
+        else:
+            import os
+
+            term_size = os.get_terminal_size()
+            print("\n", "\u2500" * (term_size.columns - 1))
```

Comment:
In another conversation, you said about the leading newline:

> Yes, again, the empty line is intentional as otherwise the table or horizontal line gets messed up.

Can you share the output?

For me locally, with the newline I see:

```python
Testing: 100%|███████████████████████████████████████| 1/1 [00:00<00:00, 59.16it/s]

X──────────────────────────────────────────────────────────────────────────────────
         Metric                DATALOADER 0       
──────────────────────────────────────────────────────────────────────────────────X
        test_loss           -0.3928866386413574   
───────────────────────────────────────────────────────────────────────────────────
```

Where I've added "X"s to indicate spaces that shouldn't be there

And without:

```python
Testing: 100%|████████████████████████████████████████| 1/1 [00:00<00:00, 85.22it/s]
───────────────────────────────────────────────────────────────────────────────────
         Metric                DATALOADER 0       
───────────────────────────────────────────────────────────────────────────────────
        test_loss           0.1391649842262268    
────────────────────────────────────────────────────────────────────────────────────
```

So still, I suggest

```suggestion
            print("\u2500" * (term_size.columns - 1))
```

Reply:
I always get an expected result when printing in the ```trainer.test()``` (without the need for additional spacing), but not when I use ```trainer.validate(verbose=True)```.

When I remove the space in the rich implementation (i.e. ```console.print(table)``` instead of ```console.print("", table)```), I get the following result (scroll also to the right to see that the line actually starts at the end of the previous line):
```
                                                                                                                            ┏
━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃   Metric   ┃    DATALOADER 0     ┃
┡━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│ valid_acc  │ -1.5097321271896362 │
│ valid_loss │ -1.5097321271896362 │
│  valid_x   │ -1.5097321271896362 │
└────────────┴─────────────────────┘
```

And for the native implementation, when I remove the space (i.e. ```print("\u2500" * term_size)``` instead of ```print("\n", "\u2500" * (term_size.columns - 1))```), I get the following result:
```
                                                                                                                                                                      ─
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
         Metric                DATALOADER 0
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
        valid_acc           0.5556910634040833
       valid_loss           0.5556910634040833
         valid_x            0.5556910634040833
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```
Note that in the native implementation to account for the additional jump, I have to use ```term_size - 1```.

Reply:
I've removed the spacing, for ```trainer.test(()``` this is completely fine, however for ```trainer.validate(verbose=True)``` I get the unexpected results as shown above. Maybe you can check whether you get something similar?

Reply:
I cannot reproduce. Looks perfect to me as it is currently for both validate and test.

Resolving, but tests should check this.